structopt-derive/src/lib.rs

```diff
@@ -449,15 +449,27 @@ fn gen_augment_clap_enum(
 ) -> TokenStream {
     use syn::Fields::*;
 
-    let subcommands = variants.iter().map(|variant| {
-        let attrs = Attrs::from_struct(
-            variant.span(),
-            &variant.attrs,
-            Name::Derived(variant.ident.clone()),
-            Some(parent_attribute),
-            parent_attribute.casing(),
-            parent_attribute.env_casing(),
-        );
+    let (flatten_variants, variants): (Vec<_>, Vec<_>) = variants.iter()
+        .map(|variant| {
+            let attrs = Attrs::from_struct(
+                variant.span(),
+                &variant.attrs,
+                Name::Derived(variant.ident.clone()),
+                Some(parent_attribute),
+                parent_attribute.casing(),
+                parent_attribute.env_casing(),
+            );
+            (variant, attrs)
+        })
+        .partition(|(_, attrs)| {
+            let kind = attrs.kind();
+            match &*kind {
+                Kind::FlattenStruct => true,
+                _ => false
+            }
+        });
```

Comment:
There's no need to separate flattened subcommands from the normal ones. I think it should look like this:
```rust
let subcommands = variants.iter().map(|variant| {
    // ...
    if let Kind::Flatten = kind {
        quote! {
            let app = augment_clap(app);
        }
    } else {
        quote! {
            let app = app.subcommand({
                // just as before
            });
        }
    }
});
```

Reply:
The same story with `from_subcommand`

Reply:
Ok that one for `gen_augment_clap_enum` is fine but it doesn't work for `gen_from_subcommand` because they are match arms.

Reply:
Ok I made it work without the match but... meh...
```rust
    fn from_subcommand<'a, 'b>(
        sub: (&'b str, Option<&'b ::structopt::clap::ArgMatches<'a>>),
    ) -> Option<Self> {
        if let Some(res) = <BaseCli as ::structopt::StructOptInternal>::from_subcommand(sub) {
            return Some(GroupCli::BaseCli(res));
        }
        if let ("command3", Some(matches)) = sub {
            return Some(GroupCli::Command3(
                <Command3 as ::structopt::StructOpt>::from_clap(matches),
            ));
        };
        None
    }
```

Reply:
Maybe I could have gone with this `if else` approach but I really don't like that flattened subcommands are prioritized over normal ones. Also... like you said, meh...

Let's back up for the partition and then something like 
```rust
match sub {
    // normal subcommands ...
    ("command3") => Some(GroupCli::Command3(
        <Command3 as ::structopt::StructOpt>::from_clap(matches),
    )),

    other => {
        // you can use if ... else if ... else if you want
        <FlattenedType as StructOpt>::from_subcommand(other)
            .or_else(|| <AnotherFlattenedType as StructOpt>::from_subcommand(other))
            .or_else(|| /* ... */)
            .map(|res| GroupCli::BaseCli(res))
    }
}
```

Reply:
done